### ISSUE TITLE
system.databases engine_full column

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#42110]
**Use case**

Replicated databases can have parameters & settings.

**Describe the solution you'd like**

Ability to access whole engine definition of database via system tables (just like system.tables table)

Another columns which make sense to have:

create_database_query,

**Describe alternatives you've considered**

SHOW CREATE DATABASE db_name

**Comment by SakshiT25** [comment]
Hi, I'm interested in working on this feature. Kindly guide me for the further steps.

**Comment by tavplubix** [comment]
> Hi, I'm interested in working on this feature. Kindly guide me for the further steps.

See https://github.com/ClickHouse/ClickHouse/blob/master/CONTRIBUTING.md and https://clickhouse.com/docs/en/development/developer-instruction/. 

You can find code that is responsible for reading from system.databases here: https://github.com/ClickHouse/ClickHouse/blob/f0e193d5a51725f524945adf55e85b80413f9c77/src/Storages/System/StorageSystemDatabases.cpp. `IDatabase::getCreateDatabaseQuery` method also will be useful (engine definition can be extracted from that CREATE query and serialized with `serializeAST` function). Feel free to ask questions if you need any help.

**Comment by ddelnano** [comment]
@tavplubix I was looking into this as well (happy to let @SakshiT25 take the lead though!). I'm running into issues building clickhouse based on the instructions you linked to and was hoping you could help.

When I run `ninja` I'm receiving the following error and it seems the `protoc` binary is not an ELF executable.
```
vagrant@ubuntu-jammy:/vagrant/build$ ninja
[0/2] Re-checking globbed directories...
[3237/10850] Generating orc_proto.pb.h, orc_proto.pb.cc
FAILED: contrib/arrow-cmake/orc_proto.pb.h contrib/arrow-cmake/orc_proto.pb.cc /vagrant/build/contrib/arrow-cmake/orc_proto.pb.h /vagrant/build/contrib/arrow-cmake/orc_proto.pb.cc
cd /vagrant/build/contrib/arrow-cmake && /vagrant/build/contrib/protobuf-cmake/protoc -I /vagrant/contrib/orc/c++/../proto --cpp_out="/vagrant/build/contrib/arrow-cmake" /vagrant/contrib/orc/c++/../proto/orc_proto.proto
/bin/sh: 1: /vagrant/build/contrib/protobuf-cmake/protoc: Exec format error
[3239/10850] Building C object contrib/openldap-cmake/CMakeFiles/_ldap_r.dir/__/openldap/libraries/libldap/sbind.c.o
ninja: build stopped: subcommand failed.

vagrant@ubuntu-jammy:/vagrant/build$ file ./contrib/protobuf-cmake/protoc
./contrib/protobuf-cmake/protoc: data
```

Any idea on what this could be from? What's the best place for getting support for build issues?

**Comment by SakshiT25** [comment]
Hey, I had a small doubt, can I set up the environment on mac without downloading VM and ubuntu.. or is it necessary?

**Comment by SakshiT25** [comment]
@tavplubix Could you please help me to know the exact requirement? I've started looking into the issue. Currently performing the setup process. 

- As per the use case definition, I understand that we are looking for database engine definitions. Just like system tables, could you help with the main file to which it needs to be added?
- Could you help me understand the code structure? I went through databases files and table files for reference, we are reading the databases from StorageSystemDatabases.cpp, I want to understand the intent behind this task. What exactly we are looking for? and which file do we need to implement that change?

**Comment by SakshiT25** [comment]
Hello @tavplubix this is a follow-up for some doubts I'm having with the requirements. Kindly update me with the date and time so that we can connect and get ahead on this issue. 
Also, would like to know if is there a backing to connect to.
Thank you.
Regards,
Sakshi

**Comment by SakshiT25** [comment]
@UnamedRus Could you please help me to know the exact requirement? I've started looking into the issue. Currently performing the setup process.

As per the use case definition, I understand that we are looking for database engine definitions. Just like system tables, could you help with the main file to which it needs to be added?
Also, could you help me to understand the code structure? I went through databases files and table files for reference, we are reading the databases from StorageSystemDatabases.cpp, I want to understand the intent behind this task. What exactly we are looking for? and which file do we need to implement that change?
Is there any medium(slack or google meet) where we can connect so that I can have a clear understanding of the feature?

**Comment by tavplubix** [comment]
> Any idea on what this could be from? What's the best place for getting support for build issues?

@ddelnano, probably `protoc` was accidentally built for wrong target architecture. It may be a result of incorrectly specified build parameters or a bug in out cmake files. What is your environment? How exactly do you run cmake (which of the instructions did you follow (we have a few ones for different environments), did you specify extra flags/parameters)? Do you use cross-compilation (like building on Linux for Mac)? Please create another issue following [the template](https://github.com/ClickHouse/ClickHouse/blob/master/.github/ISSUE_TEMPLATE/50_build-issue.md)

**Comment by tavplubix** [comment]
@SakshiT25, each database in ClickHouse has "engine" which is specified in database create statement, for example:
```
dell9510 :) show create database default

┌─statement──────────────────────────────┐
│ CREATE DATABASE default
ENGINE = Atomic │
└────────────────────────────────────────┘
```
In this example database has "Atomic" engine and it is shown in `engine` column of `system.databases` table:
```
dell9510 :) select name, engine from system.databases where name='default'

┌─name────┬─engine─┐
│ default │ Atomic │
└─────────┴────────┘
```
But some engines may have arguments and settings, but arguments and settings are not shown in the `engine` column, for example:
```
dell9510 :) create database example_db_with_args engine=Replicated('/test/example_args', 'shard1', 'replica1')

Ok.

dell9510 :) select name, engine from system.databases where name='example_db_with_args'

┌─name─────────────────┬─engine─────┐
│ example_db_with_args │ Replicated │
└──────────────────────┴────────────┘

```

The purpose of this task is to add another column to `system.databases` table that will show database engine with arguments and settings (if any) in a way similar to `system.tables`:
```
dell9510 :) select database, name, engine, engine_full from system.tables where name='test_table'

┌─database─┬─name───────┬─engine──────────────┬─engine_full───────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ default  │ test_table │ ReplicatedMergeTree │ ReplicatedMergeTree('/test/test_table', 'replica1') ORDER BY n SETTINGS index_granularity = 8192                      │
│ test     │ test_table │ ReplicatedMergeTree │ ReplicatedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}') ORDER BY date SETTINGS index_granularity = 8192 │
└──────────┴────────────┴─────────────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

To implement this you need to add the new column in `StorageSystemDatabases::getNamesAndTypes` and fill it in `StorageSystemDatabases::fillData`. You can get `ASTCreateQuery` for a database using `IDatabase::getCreateDatabaseQuery`, extract `ASTCreateQuery::storage` (which represents full engine definition) and serialize it into string using `serializeAST`.

**Comment by SakshiT25** [comment]
@tavplubix  Hello, I understood the requirements. Thank you. 
Currently, I'm facing an error while updating submodules in a .icu file. I've tried debugging it many times but it is showing again.
Kindly help me with the issue.
<img width="571" alt="Screenshot 2022-10-18 at 7 42 10 AM" src="https://user-images.githubusercontent.com/65718952/196859696-e3298659-9c2d-4d96-b4f9-d01530d39d38.png">

**Comment by SakshiT25** [comment]
I first generated and added the ssh key for the GitHub access. then using 

> git clone --recursive <copied the link of ssh from code tab>

> cd Clickhouse

> git submodule update //for updating the submodules after running the init command.

**Comment by tavplubix** [comment]
You need to install `git-lfs`: https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage?platform=mac

**Comment by SakshiT25** [comment]
<img width="1440" alt="lsb_issue" src="https://user-images.githubusercontent.com/65718952/200226641-3d9e2e44-56c9-4305-84bd-c924821304a4.png">
@tavplubix 
Good morning Sir, Could you help me in resolving the error by running the following command?
sudo bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)"

I installed wget on mac. but now it's showing an error about "lsb_release' add-apt-repository gpg" because it might not support mac OSX. Is there any alternative command I can try or should I download the clang compiler directly?

**Comment by tavplubix** [comment]
Sorry, but I'm not going to help with this, it's not related to ClickHouse. Please use Google and StackOverflow.

**Comment by SakshiT25** [comment]
Hello, @tavplubix  Good evening, the last error I tried resolving but could not do on MAC OSX. Therefore, I downloaded VM and installed ubuntu. However, I'm getting errors while installing the C++ compiler.
Could you help me with this? I googled StackOverflow and various other websites and tried the following commands. 
sudo apt-get install ca-certificates
sudo apt upgrade.
![c++ error](https://user-images.githubusercontent.com/65718952/201668532-a16d449a-5e05-4c35-9800-b7243cb8ecab.jpg)

**Comment by SakshiT25** [comment]
@tavplubix Also while building ninja I'm getting an error in proto while linking.
![error iin protoc](https://user-images.githubusercontent.com/65718952/201670831-6c885768-580c-4266-a83b-e44261d2c8dd.jpg)
 file.

**Comment by lingtaolf** [comment]
Hello, I can take over if no one is still working on it.

**Comment by UnamedRus** [comment]
Merged in 22.12
